# Text into Data

```yaml
Name:   Ji Hyun Kim 
Computing ID:   mqa4qu
Code Guide: M02_01_Importing-Persuasion
Purpose:    Convert a different text from raw text into a data frame of tokens and preserving its OHCO. Then extract some statistical features from the resulting corpus.
```

## Import

In [1]:
import pandas as pd
import configparser

/var/folders/4f/kdhsm9_j5lq_3f_lpx0y2vpm0000gn/T/ipykernel_40044/3503867414.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [2]:
config = configparser.ConfigParser()
config.read("../../../env.ini")

data_home = config['DEFAULT']['data_home']
output_dir = config['DEFAULT']['output_dir']

data_home, output_dir

('/Users/jihyunkim/desktop/uva_docs/DS5001/data',
 '/Users/jihyunkim/desktop/uva_docs/DS5001/output')

In [3]:
text_file = f"{data_home}/gutenberg/pg161.txt"
csv_file  = f"{output_dir}/austen-sense-and-sensibility.csv"

OHCO = ['book_num', 'chap_num', 'para_num', 'sent_num', 'token_num']

In [4]:
LINES = pd.DataFrame(open(text_file, 'r', encoding='utf-8-sig').readlines(), columns=['line_str'])
LINES.index.name = 'line_num'
LINES.line_str = LINES.line_str.str.replace(r'\n+', ' ', regex=True).str.strip()

LINES['book_num'] = 0

# LINES.head(10)

## Tasks

In [5]:
# Extract Title
title = LINES.loc[0].line_str.replace('The Project Gutenberg EBook of ', '')
print(title)

Sense and Sensibility, by Jane Austen


### Task 1: Clip the cruft
Remove Gutenberg's front and back matter using the lines that indicate the start and end of the project.

In [6]:
clip_pats = [
    r"\*\*\*\s*START OF (?:THE|THIS) PROJECT",
    r"\*\*\*\s*END OF (?:THE|THIS) PROJECT"
]

In [7]:
pat_a = LINES.line_str.str.match(clip_pats[0])
pat_b = LINES.line_str.str.match(clip_pats[1])

line_a = LINES.loc[pat_a].index[0] + 1
line_b = LINES.loc[pat_b].index[0] - 1

line_a, line_b

(20, 12666)

In [8]:
LINES = LINES.loc[line_a : line_b]
LINES.head(20)
LINES.tail(10)

line_str  book_num
line_num                                                             
12657                                                               0
12658                                                               0
12659                                                               0
12660                                                               0
12661                                                               0
12662                                                               0
12663                                                               0
12664                                                               0
12665     End of the Project Gutenberg EBook of Sense an...         0
12666                                                               0

### Task 2
Chunk by chapter, using the pattern of locating the headers in the data frame, assigning them 
numbers, forward-filling those numbers, and then grouping by number (and cleaning up).

#### Task 2 - Chunk by Chapters

In [9]:
chap_pat = r"^\s*(?:chapter|letter)\s+\d+"
chap_lines = LINES.line_str.str.match(chap_pat, case=False)
LINES.loc[chap_lines]

line_str  book_num
line_num                      
42         CHAPTER 1         0
196        CHAPTER 2         0
399        CHAPTER 3         0
561        CHAPTER 4         0
756        CHAPTER 5         0
858        CHAPTER 6         0
986        CHAPTER 7         0
1112       CHAPTER 8         0
1244       CHAPTER 9         0
1448      CHAPTER 10         0
1665      CHAPTER 11         0
1816      CHAPTER 12         0
1997      CHAPTER 13         0
2281      CHAPTER 14         0
2440      CHAPTER 15         0
2718      CHAPTER 16         0
2945      CHAPTER 17         0
3153      CHAPTER 18         0
3331      CHAPTER 19         0
3632      CHAPTER 20         0
3913      CHAPTER 21         0
4214      CHAPTER 22         0
4532      CHAPTER 23         0
4767      CHAPTER 24         0
5001      CHAPTER 25         0
5197      CHAPTER 26         0
5454      CHAPTER 27         0
5732      CHAPTER 28         0
5883      CHAPTER 29         0
6324      CHAPTER 30         0
6628      CHAPTER 31         0
7004      CHAPTER 32         0
7278      CHAPTER 33         0
7601      CHAPTER 34         0
7888      CHAPTER 35         0
8152      CHAPTER 36         0
8456      CHAPTER 37         0
8900      CHAPTER 38         0
9205      CHAPTER 39         0
9408      CHAPTER 40         0
9706      CHAPTER 41         0
9977      CHAPTER 42         0
10155     CHAPTER 43         0
10490     CHAPTER 44         0
11060     CHAPTER 45         0
11278     CHAPTER 46         0
11571     CHAPTER 47         0
11838     CHAPTER 48         0
11986     CHAPTER 49         0
12410     CHAPTER 50         0

#### Task 2 - Assign numbers

In [10]:
LINES.loc[chap_lines, 'chap_num'] = [i+1 for i in range(LINES.loc[chap_lines].shape[0])]
LINES.loc[chap_lines]

line_str  book_num  chap_num
line_num                                
42         CHAPTER 1         0       1.0
196        CHAPTER 2         0       2.0
399        CHAPTER 3         0       3.0
561        CHAPTER 4         0       4.0
756        CHAPTER 5         0       5.0
858        CHAPTER 6         0       6.0
986        CHAPTER 7         0       7.0
1112       CHAPTER 8         0       8.0
1244       CHAPTER 9         0       9.0
1448      CHAPTER 10         0      10.0
1665      CHAPTER 11         0      11.0
1816      CHAPTER 12         0      12.0
1997      CHAPTER 13         0      13.0
2281      CHAPTER 14         0      14.0
2440      CHAPTER 15         0      15.0
2718      CHAPTER 16         0      16.0
2945      CHAPTER 17         0      17.0
3153      CHAPTER 18         0      18.0
3331      CHAPTER 19         0      19.0
3632      CHAPTER 20         0      20.0
3913      CHAPTER 21         0      21.0
4214      CHAPTER 22         0      22.0
4532      CHAPTER 23         0      23.0
4767      CHAPTER 24         0      24.0
5001      CHAPTER 25         0      25.0
5197      CHAPTER 26         0      26.0
5454      CHAPTER 27         0      27.0
5732      CHAPTER 28         0      28.0
5883      CHAPTER 29         0      29.0
6324      CHAPTER 30         0      30.0
6628      CHAPTER 31         0      31.0
7004      CHAPTER 32         0      32.0
7278      CHAPTER 33         0      33.0
7601      CHAPTER 34         0      34.0
7888      CHAPTER 35         0      35.0
8152      CHAPTER 36         0      36.0
8456      CHAPTER 37         0      37.0
8900      CHAPTER 38         0      38.0
9205      CHAPTER 39         0      39.0
9408      CHAPTER 40         0      40.0
9706      CHAPTER 41         0      41.0
9977      CHAPTER 42         0      42.0
10155     CHAPTER 43         0      43.0
10490     CHAPTER 44         0      44.0
11060     CHAPTER 45         0      45.0
11278     CHAPTER 46         0      46.0
11571     CHAPTER 47         0      47.0
11838     CHAPTER 48         0      48.0
11986     CHAPTER 49         0      49.0
12410     CHAPTER 50         0      50.0

#### Task 2 - Forward Filling chapter numbers

In [11]:
LINES.chap_num = LINES.chap_num.ffill()
# LINES.sample(10)
LINES.head(50)

line_str  book_num  \
line_num                                                                
20                                                                  0   
21                                                                  0   
22                                                                  0   
23                                                                  0   
24                                                                  0   
25                                                                  0   
26                                                                  0   
27                                                                  0   
28                                                                  0   
29                                                                  0   
30                                                                  0   
31                                                                  0   
32                                                                  0   
33                                    SENSE AND SENSIBILITY         0   
34                                                                  0   
35                                           by Jane Austen         0   
36                                                                  0   
37                                                   (1811)         0   
38                                                                  0   
39                                                                  0   
40                                                                  0   
41                                                                  0   
42                                                CHAPTER 1         0   
43                                                                  0   
44                                                                  0   
45        The family of Dashwood had long been settled i...         0   
46        was large, and their residence was at Norland ...         0   
47        their property, where, for many generations, t...         0   
48        respectable a manner as to engage the general ...         0   
49        surrounding acquaintance.  The late owner of t...         0   
50        man, who lived to a very advanced age, and who...         0   
51        life, had a constant companion and housekeeper...         0   
52        death, which happened ten years before his own...         0   
53        alteration in his home; for to supply her loss...         0   
54        into his house the family of his nephew Mr. He...         0   
55        inheritor of the Norland estate, and the perso...         0   
56        bequeath it.  In the society of his nephew and...         0   
57        children, the old Gentleman's days were comfor...         0   
58        attachment to them all increased.  The constan...         0   
59        Mrs. Henry Dashwood to his wishes, which proce...         0   
60        interest, but from goodness of heart, gave him...         0   
61        comfort which his age could receive; and the c...         0   
62                children added a relish to his existence.         0   
63                                                                  0   
64        By a former marriage, Mr. Henry Dashwood had o...         0   
65        lady, three daughters.  The son, a steady resp...         0   
66        amply provided for by the fortune of his mothe...         0   
67        and half of which devolved on him on his comin...         0   
68        marriage, likewise, which happened soon afterw...         0   
69        wealth.  To him therefore the succession to th...         0   

          chap_num  
line_num            
20             NaN  
21             NaN  
22             NaN  
23             NaN  
24             NaN  
25             NaN  
26             NaN  
27             NaN  
28             NaN  
29             NaN  
30

#### Task 2 - Clean up

In [12]:
LINES = LINES.dropna(subset=['chap_num'])
LINES = LINES.loc[~chap_lines]
LINES.chap_num = LINES.chap_num.astype('int')
LINES.head(20)

line_str  book_num  \
line_num                                                                
43                                                                  0   
44                                                                  0   
45        The family of Dashwood had long been settled i...         0   
46        was large, and their residence was at Norland ...         0   
47        their property, where, for many generations, t...         0   
48        respectable a manner as to engage the general ...         0   
49        surrounding acquaintance.  The late owner of t...         0   
50        man, who lived to a very advanced age, and who...         0   
51        life, had a constant companion and housekeeper...         0   
52        death, which happened ten years before his own...         0   
53        alteration in his home; for to supply her loss...         0   
54        into his house the family of his nephew Mr. He...         0   
55        inheritor of the Norland estate, and the perso...         0   
56        bequeath it.  In the society of his nephew and...         0   
57        children, the old Gentleman's days were comfor...         0   
58        attachment to them all increased.  The constan...         0   
59        Mrs. Henry Dashwood to his wishes, which proce...         0   
60        interest, but from goodness of heart, gave him...         0   
61        comfort which his age could receive; and the c...         0   
62                children added a relish to his existence.         0   

          chap_num  
line_num            
43               1  
44               1  
45               1  
46               1  
47               1  
48               1  
49               1  
50               1  
51               1  
52               1  
53               1  
54               1  
55               1  
56               1  
57               1  
58               1  
59               1  
60               1  
61               1  
62               1

#### Task 2 - Group by Number

In [13]:
OHCO[:2]

CHAPS = LINES.groupby(OHCO[:2])\
    .line_str.apply(lambda x: '\n'.join(x))\
    .to_frame('chap_str')

CHAPS['chap_str'] = CHAPS.chap_str.str.strip()
CHAPS.head(10)

chap_str
book_num chap_num                                                   
0        1         The family of Dashwood had long been settled i...
         2         Mrs. John Dashwood now installed herself mistr...
         3         Mrs. Dashwood remained at Norland several mont...
         4         "What a pity it is, Elinor," said Marianne, "t...
         5         No sooner was her answer dispatched, than Mrs....
         6         The first part of their journey was performed ...
         7         Barton Park was about half a mile from the cot...
         8         Mrs. Jennings was a widow with an ample jointu...
         9         The Dashwoods were now settled at Barton with ...
         10        Marianne's preserver, as Margaret, with more e...

### Task 3: Split Chapters into Paragraphs
Split resulting data frame into paragraphs using the regex provided

In [14]:
OHCO[:3]

['book_num', 'chap_num', 'para_num']

In [15]:
para_pat = r'\n\n+'

PARAS = CHAPS['chap_str'].str.split(para_pat, expand=True).stack()\
    .to_frame('para_str').sort_index()
PARAS.index.names = OHCO[:3]

PARAS.head(10)

para_str
book_num chap_num para_num                                                   
0        1        0         The family of Dashwood had long been settled i...
                  1         By a former marriage, Mr. Henry Dashwood had o...
                  2         The old gentleman died: his will was read, and...
                  3         Mr. Dashwood's disappointment was, at first, s...
                  4         His son was sent for as soon as his danger was...
                  5         Mr. John Dashwood had not the strong feelings ...
                  6         He was not an ill-disposed young man, unless t...
                  7         When he gave his promise to his father, he med...
                  8         No sooner was his father's funeral over, than ...
                  9         So acutely did Mrs. Dashwood feel this ungraci...

In [16]:
PARAS['para_str'] = PARAS['para_str'].str.replace(r'\n', ' ', regex=True)
PARAS['para_str'] = PARAS['para_str'].str.strip()
PARAS = PARAS[~PARAS['para_str'].str.match(r'^\s*$')]

PARAS.head(10)

para_str
book_num chap_num para_num                                                   
0        1        0         The family of Dashwood had long been settled i...
                  1         By a former marriage, Mr. Henry Dashwood had o...
                  2         The old gentleman died: his will was read, and...
                  3         Mr. Dashwood's disappointment was, at first, s...
                  4         His son was sent for as soon as his danger was...
                  5         Mr. John Dashwood had not the strong feelings ...
                  6         He was not an ill-disposed young man, unless t...
                  7         When he gave his promise to his father, he med...
                  8         No sooner was his father's funeral over, than ...
                  9         So acutely did Mrs. Dashwood feel this ungraci...

### Task 4: Split paragraphs into sentences
Split resulting data frame into sentences using the regex provided.

In [17]:
sent_pat = r'[.?!;:]+'
SENTS = PARAS['para_str'].str.split(sent_pat, expand=True).stack()\
    .to_frame('sent_str')
SENTS.index.names = OHCO[:4]

In [18]:
SENTS = SENTS[~SENTS['sent_str'].str.match(r'^\s*$')]
SENTS.sent_str = SENTS.sent_str.str.strip()

SENTS.head(10)
# SENTS.sample(10)

sent_str
book_num chap_num para_num sent_num                                                   
0        1        0        0         The family of Dashwood had long been settled i...
                           1         Their estate was large, and their residence wa...
                           2         The late owner of this estate was a single man...
                           3         But her death, which happened ten years before...
                           4         for to supply her loss, he invited and receive...
                           5         Henry Dashwood, the legal inheritor of the Nor...
                           6         In the society of his nephew and niece, and th...
                           7                      His attachment to them all increased
                           8                              The constant attention of Mr
                           9                                                   and Mrs

### Task 4: Split sentences into tokens
Split resulting data frame into tokens using the regex provided.

In [19]:
token_pat = r"[\s',-]+"
TOKENS = SENTS['sent_str'].str.split(token_pat, expand=True).stack()\
    .to_frame('token_str')

TOKENS.index.names = OHCO[:5]

# TOKENS.head(10)
TOKENS.sample(10)
TOKENS.head(10)

token_str
book_num chap_num para_num sent_num token_num          
0        1        0        0        0               The
                                    1            family
                                    2                of
                                    3          Dashwood
                                    4               had
                                    5              long
                                    6              been
                                    7           settled
                                    8                in
                                    9            Sussex

### Task 5: Combine Persuasion
Combine both Persuasion and Sense and Sensibility into a single data frame with an appropriately modified OHCO list. In other words, make sure your index includes a new index level for the book. Use the attached CSV to get the Persuasion data and then import it into your notebook as a data frame.

In [20]:
csv_file2  = f"{output_dir}/austen-persuasion.csv"
persuasion_df = pd.read_csv(csv_file2)

persuasion_df.drop('term_str', axis=1, inplace=True)
persuasion_df.insert(0, 'book_num', 1)
# persuasion_df.sample(30)

full_df = pd.concat([TOKENS, persuasion_df.set_index(['book_num', 'chap_num', 'para_num', 'sent_num', 'token_num'])])
full_df.sort_index(inplace=True)
# full_df.sample(30)
# full_df.head(20)
full_df

token_str
book_num chap_num para_num sent_num token_num            
0        1        0        0        0                 The
                                    1              family
                                    2                  of
                                    3            Dashwood
                                    4                 had
...                                                   ...
1        24       13       0        6                  of
                                    7          Persuasion
                                    8                  by
                                    9                Jane
                                    10             Austen

[207896 rows x 1 columns]

### Task 6: Extract Vocabulary
From the combined data frame, extract a vocabulary, i.e. a data frame with term string as index, along with term frequency and term length as features.

In [21]:
full_df['term_str'] = full_df.token_str.replace(r'[\W_]+', '', regex=True).str.lower()
VOCAB = full_df.term_str.value_counts().to_frame('n').reset_index().rename(columns={'index':'term_str'})
VOCAB.index.name = 'term_id'

VOCAB['term_len'] = VOCAB.term_str.str.len()
VOCAB

term_str     n  term_len
term_id                               
0                  the  7435         3
1                   to  6923         2
2                  and  6290         3
3                   of  6146         2
4                  her  3747         3
...                ...   ...       ...
8234     unconquerable     1        13
8235          outgrown     1         8
8236      prosperously     1        12
8237           nominal     1         7
8238             finis     1         5

[8239 rows x 3 columns]

In [22]:
total_raw_tokens = VOCAB['n'].sum()
total_raw_tokens

207786

### Task 7: 
Answer the following questions by extracting features from the corpus.
1. How many raw tokens are in the combined data frame?

   <b>There are 207786 tokens</b>

2. How many distinct terms are there in the combined data frame (i.e. how big is the vocabulary)?

    8239 rows, so <b>8239 terms</b>

#### Gathering by Content Object

In [23]:
def gather(ohco_level):
    global full_df
    level_name = OHCO[ohco_level-1].split('_')[0]
    df = full_df.groupby(OHCO[:ohco_level])\
        .token_str.apply(lambda x: x.str.cat(sep=' '))\
        .to_frame(f"{level_name}_str")
    return df

In [24]:
book_level = gather(1)

count_terms_book0 = set(book_level.loc[0, 'book_str'].split())
count_terms_book1 = set(book_level.loc[1, 'book_str'].split())

more_terms = len(count_terms_book0) - len(count_terms_book1)
more_terms

665

### Task 7: 
3. How many more terms does the vocabulary of Sense and Sensibility have than that of Persuasion?

    There are <b>665</b> more terms in Sense and Sensibility

In [25]:
count_token_chap = full_df.groupby(['book_num', 'chap_num']).size()
avg_tokens_chap = count_token_chap.mean().round().astype(int)
avg_tokens_chap

2809

### Task 7: 
4. What is the average number of tokens, rounded to an integer, per chapter in the corpus?

    There are <b>2809</b> number of tokens per chapter in the corpus

In [26]:
count_token_par = full_df.groupby(['book_num', 'chap_num', 'para_num']).size()
avg_tokens_par = count_token_par.mean().round().astype(int)
avg_tokens_par

74

### Task 7: 
5. What is the average number of tokens, rounded to an integer, per paragraph in the corpus?

    There are <b>74</b> number of tokens per paragraph in the corpus